sdks/java/core/src/main/java/org/apache/beam/sdk/util/ReduceFnRunner.java

```diff
@@ -936,16 +936,17 @@ private void cancelEndOfWindowAndGarbageCollectionTimers(
   }
 
   /**
-   * Return when {@code window} should be garbage collected. If the window is the GlobalWindow,
-   * that will be the end of the window. Otherwise, add the allowed lateness to the end of
-   * the window.
+   * Return when {@code window} should be garbage collected. If the window's expiration time is on
+   * or after the end of the global window, it will be truncated to the end of the global window.
    */
   private Instant garbageCollectionTime(W window) {
-    Instant maxTimestamp = window.maxTimestamp();
-    if (maxTimestamp.isBefore(GlobalWindow.INSTANCE.maxTimestamp())) {
-      return maxTimestamp.plus(windowingStrategy.getAllowedLateness());
+    if (window
```

Comment:
I think my code avoids overflow pre-emptively for non-negative allowed lateness, no? In Joda overflow results in an arithmetic exception, not wraparound, so I thought this was slightly more readable (and maybe less expensive, though that probably doesn't matter here)